File: 25.Reverse Nodes in k-Group/ReverseNodesinkGroup.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void swap(ListNode *front,ListNode *rear){
        ListNode * tmp=front->next; 
        ListNode * next=NULL;
        while(front!=rear)
        {
            next=tmp->next;
            tmp->next=front;
            front=tmp;
            tmp=next;
        }
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(head == NULL)
            return head;
        ListNode myhead(-1);
        myhead.next=head;
        ListNode* rear=head;
        for(int i=1;i<k && rear!=NULL;i++)
            rear=rear->next;
        if(rear==NULL)
            return head;
        ListNode* prev= &myhead;
        ListNode* next=rear->next;
        ListNode* front=head;
        while(rear!=NULL){
            swap(front,rear);
            prev->next=rear;
            front->next=next;
            prev=front;
            front=next;
            rear=next;
            for(int i=1;i<k && rear!=NULL;i++)
                rear=rear->next;
            if(rear!=NULL)
                next=rear->next;
        }
        return myhead.next;
    }
};

File: 25.Reverse Nodes in k-Group/ReverseNodesink-Group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        
        if(head == null)
            return head;
        //ListNode dummy = new ListNode(-1);
        //dummy.next = head;
        ListNode current = head;
        int count = 0;
        while(count != k && current != null)
        {
            current = current.next;
            count ++;
        }
        if(count == k)
        {
            current = reverseKGroup(current, k);
            while(count -- > 0)
            {
                ListNode temp = head.next;
                head.next = current;
                current = head;
                head = temp;
            }
            head = current;
        }
        return head;
    }
}